### PR TITLE
SUS-1563: Introduce maintenance script to list and delete duplicate blocks from ipblocks table and perform schema update

### PR DESCRIPTION
**Pull request by mszabo-wikia**
https://wikia-inc.atlassian.net/browse/SUS-1563

### CODE REVIEW

File: maintenance/wikia/updateIpblocksSchema.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		}
+	}
+
+	/**
+	 * Load and print block IDs to be deleted
+	 * @return array list of block IDs to be deleted and count of block IDs to be deleted
+	 */
+	private function getDuplicateBlocks(): array {
+		/*
+			SELECT GROUP_CONCAT(ipb_id) FROM ipblocks GROUP BY ipb_address, ipb_user HAVING count(*) > 1;
```
Comment (WikiaTech) [inline maintenance/wikia/updateIpblocksSchema.php]: MAJOR Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: maintenance/wikia/updateIpblocksSchema.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	/**
+	 * Load and print block IDs to be deleted
+	 * @return array list of block IDs to be deleted and count of block IDs to be deleted
+	 */
+	private function getDuplicateBlocks(): array {
+		/*
+			SELECT GROUP_CONCAT(ipb_id) FROM ipblocks WHERE ipb_auto = 0 GROUP BY ipb_address, ipb_user HAVING count(*) > 1;
```
Comment (WikiaTech) [inline maintenance/wikia/updateIpblocksSchema.php:35]: MAJOR Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)

**Review by macbre — APPROVED** [review]

**Review by macbre — APPROVED** [review]

File: maintenance/wikia/updateIpblocksSchema.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			$this->output( "target: {$row->ipb_address}, duplicates: {$row->duplicate_ids}, preserve: $newest\n" );
+
+			$allIds = array_merge( $allIds, $duplicates );
+		}
+
+		$allIdsCount = count( $allIds );
+		if ( $allIdsCount > 0 ) {
+			$this->output( "{$this->wiki}: $allIdsCount block IDs to be deleted.\n" );
+		} else {
+			$this->output( "{$this->wiki}: nothing to do" );
```
Comment (ghost) [inline maintenance/wikia/updateIpblocksSchema.php]: Please, add a proper line ending here.

File: maintenance/wikia/updateIpblocksSchema.php (hunk 2 above)
Comment (mszabo-wikia) [reply maintenance/wikia/updateIpblocksSchema.php:35]: this is not commented out code

File: includes/db/Database.php [hunk 4]
```diff
@@ -2671,7 +2671,7 @@ function lowPriorityOption() {
 	 * @throws DBUnexpectedError
 	 * @throws MWException
 	 */
-	function delete( $table, $conds, $fname = 'DatabaseBase::delete' ) {
+	function delete( string $table, $conds, $fname = 'DatabaseBase::delete' ) {
```
Comment (macbre) [inline includes/db/Database.php:2674]: :+1:

**Review by macbre — APPROVED** [review]

**Comment by WikiaTech** [comment]
SonarQube analysis reported 83 issues
* ![MAJOR][MAJOR] 40 major
* ![MINOR][MINOR] 19 minor
* ![INFO][INFO] 24 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Database.php](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php): File "Database.php" has 3795 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS104)
1. ![MAJOR][MAJOR] [Database.php#L202](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L202): Class "DatabaseBase" has 157 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
1. ![MAJOR][MAJOR] [Database.php#L592](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L592): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)
1. ![MAJOR][MAJOR] [Database.php#L858](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L858): The Cyclomatic Complexity of this function "query" is 33 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [Database.php#L858](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L858): This function "query" has 154 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)
1. ![MAJOR][MAJOR] [Database.php#L903](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L903): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Database.php#L914](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L914): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Database.php#L951](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L951): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1067)
1. ![MAJOR][MAJOR] [Database.php#L957](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L957): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [Database.php#L1146](https://github.com/Wikia/app/blob/3dd007e0a3dbf61be0a85a6ec0b8508eea5b4719/includes/db/Database.php#L1146): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 83 issues
* ![MAJOR][MAJOR] 40 major
* ![MINOR][MINOR] 19 minor
* ![INFO][INFO] 24 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Database.php](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php): File "Database.php" has 3795 lines, which is greater than 1000 authorized. Split it into smaller files. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS104)
1. ![MAJOR][MAJOR] [Database.php#L202](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L202): Class "DatabaseBase" has 157 methods, which is greater than 20 authorized. Split it into smaller classes. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1448)
1. ![MAJOR][MAJOR] [Database.php#L592](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L592): Take the required action to fix the issue indicated by this "FIXME" comment. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1134)
1. ![MAJOR][MAJOR] [Database.php#L858](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L858): The Cyclomatic Complexity of this function "query" is 33 which is greater than 20 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1541)
1. ![MAJOR][MAJOR] [Database.php#L858](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L858): This function "query" has 154 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS138)
1. ![MAJOR][MAJOR] [Database.php#L903](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L903): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Database.php#L914](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L914): Refactor this code to rely on RequestContext instead of context-dependent globals. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)
1. ![MAJOR][MAJOR] [Database.php#L951](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L951): Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1067)
1. ![MAJOR][MAJOR] [Database.php#L957](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L957): Add curly braces around the nested statement(s). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS121)
1. ![MAJOR][MAJOR] [Database.php#L1146](https://github.com/Wikia/app/blob/8b1dd4bf619020f03d9489a0bab550a02323fbfe/includes/db/Database.php#L1146): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**